gluonnlp/model/beam_search.py

```diff
@@ -124,14 +150,37 @@ def _choose_states(F, states, indices):
         Each NDArray/Symbol should have shape (N, ...).
     """
     if isinstance(states, list):
-        return [_choose_states(F, ele, indices) for ele in states]
+        assert not state_info or isinstance(state_info, list), \
+                'state and state_info doesn\'t match'
+        if not state_info:
+            state_info = [None] * len(states)
+        return [_choose_states(F, d, s, indices) for d, s in zip(states, state_info)]
     elif isinstance(states, tuple):
-        return tuple(_choose_states(F, ele, indices) for ele in states)
+        assert not state_info or isinstance(state_info, tuple), \
+                'state and state_info doesn\'t match'
+        if not state_info:
+            state_info = [None] * len(states)
+            state_info = tuple(state_info)
+        return tuple(_choose_states(F, d, s, indices) for d, s in zip(states, state_info))
     elif isinstance(states, dict):
-        return {k: _choose_states(F, v, indices)
+        assert not state_info or isinstance(state_info, dict), \
+                'state and state_info doesn\'t match'
+        if not state_info:
+            state_info = {k: None for k in states.keys()}
+        return {k: _choose_states(F, v, state_info[k], indices)
                 for k, v in states.items()}
     elif isinstance(states, (mx.nd.NDArray, mx.sym.Symbol)):
-        return F.take(states, indices)
+        assert not state_info or isinstance(state_info, dict), \
+                'state and state_info doesn\'t match'
+        if not state_info:
+            state_info = {'__layout__': 'NC'}
+        batch_axis = state_info['__layout__'].find('N')
+        if batch_axis != 0:
+            states = states.swapaxes(0, batch_axis)
+        states = F.take(states, indices)
+        if batch_axis != 0:
+            states = states.swapaxes(0, batch_axis)
+        return states
```

Comment:
So the returned states always have batch in the first dim even it is not the case for the input? Would it cause some inconsistency?

Reply:
Because of the two swaps, the batch dimension should be where it used to be.

Reply:
Yes, you are right.

Reply:
We can directly use take(axis=batch_axis) once https://github.com/apache/incubator-mxnet/pull/11326 is merged.